File: server/datasets/python_source-code/Multiple-D1/organic_2.py
"""
Quick Sort - Organic Submission #2
Hoare partition scheme with two converging pointers.
"""

def hoare_partition(arr, low, high):
    pivot = arr[low]
    i = low - 1
    j = high + 1
    while True:
        i += 1
        while arr[i] < pivot:
            i += 1
        j -= 1
        while arr[j] > pivot:
            j -= 1
        if i >= j:
            return j
        arr[i], arr[j] = arr[j], arr[i]

def quick_sort_hoare(arr, low, high):
    if low < high:
        p = hoare_partition(arr, low, high)
        quick_sort_hoare(arr, low, p)
        quick_sort_hoare(arr, p + 1, high)

if __name__ == "__main__":
    nums = [19, 22, 63, 105, 2, 46]
    quick_sort_hoare(nums, 0, len(nums) - 1)
    print(nums)

File: server/datasets/java_source-code/Multiple-D1/organic_2.java
/*
 * Quick Sort - Organic Submission #2
 * Hoare partition scheme with two converging pointers.
 */

class Solution {
    public static int hoarePartition(int[] arr, int low, int high) {
        int pivot = arr[low];
        int i = low - 1;
        int j = high + 1;
        while (true) {
            do { i++; } while (arr[i] < pivot);
            do { j--; } while (arr[j] > pivot);
            if (i >= j) return j;
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    public static void quickSortHoare(int[] arr, int low, int high) {
        if (low < high) {
            int p = hoarePartition(arr, low, high);
            quickSortHoare(arr, low, p);
            quickSortHoare(arr, p + 1, high);
        }
    }

    public static void main(String[] args) {
        int[] nums = {19, 22, 63, 105, 2, 46};
        quickSortHoare(nums, 0, nums.length - 1);
        System.out.println(java.util.Arrays.toString(nums));
    }
}
